sympy/matrices/dense.py

```diff
@@ -880,7 +880,10 @@ def row_del(self, i):
         row
         col_del
         """
-        self._mat = self._mat[:i*self.cols] + self._mat[(i + 1)*self.cols:]
+        if i >= -self.rows and i < self.rows:
+            del self._mat[i*self.cols:(i + 1)*self.cols]
+        else:
+            raise IndexError("Index out of range: 'i = %s', valid -%s <= i < %s" % (i, self.rows, self.rows))
```

Comment:
I tend to use this idiom:

```
if i < -self.rows or i >= self.rows:
    raise ...
del self._mat[...]
```

The advantage is that error checking and error message are kept nearer together.
Also, if additional checks are required, I can insert them as additional `if` clauses without having to touch the existing logic.

One other thing: You're replacing a construction of a new matrix with an in-place update.
This could be a breaking change, if there's any code that shares the `_mat` value with another `Matrix*` object.
Unless you have excellent reasons to replace the assignment with a `del` statement, please stick with the original code.